### PR TITLE
WIP: Attitude message cleanup

### PR DESCRIPTION
**Pull request by tumbili**
This removes out all the irrelevant and redundant data out of the attitude topic.
The main reason why we do this is because the new logger app #4282 will log entire topics and so we don't want to burn the logging bandwidth on redundant data.

This needs review, testing in SITL and testing outdoors.
At several locations code has been commented out. If we agree on the content of the attitude message then this code can be removed.

### CODE REVIEW

**Comment by tumbili** [comment]
@kd0aij Review would be nice. Most of the cases are very simple. We either need euler angles or a rotation matrix so we first create the quaternion and convert.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -1746,15 +1746,16 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
...
+		matrix::Euler<float> euler(q);
 		hil_global_pos.timestamp = timestamp;
 		hil_global_pos.lat = hil_state.lat / ((double)1e7);
 		hil_global_pos.lon = hil_state.lon / ((double)1e7);
 		hil_global_pos.alt = hil_state.alt / 1000.0f;
 		hil_global_pos.vel_n = hil_state.vx / 100.0f;
 		hil_global_pos.vel_e = hil_state.vy / 100.0f;
 		hil_global_pos.vel_d = hil_state.vz / 100.0f;
-		hil_global_pos.yaw = hil_attitude.yaw;
+		hil_global_pos.yaw = euler(1);
```
Comment (kd0aij) [inline src/modules/mavlink/mavlink_receiver.cpp]: should this be euler(2) ?

File: src/modules/position_estimator_inav/position_estimator_inav_main.cpp [hunk 2]
```diff
@@ -1323,7 +1329,8 @@ int position_estimator_inav_thread_main(int argc, char *argv[])
 			local_pos.vy = y_est[1];
 			local_pos.z = z_est[0];
 			local_pos.vz = z_est[1];
-			local_pos.yaw = att.yaw;
+			matrix::Euler<float> euler(R);
+			local_pos.yaw = euler(0);
```
Comment (kd0aij) [inline src/modules/position_estimator_inav/position_estimator_inav_main.cpp]: should be euler(2) ?

File: src/modules/sdlog2/sdlog2.c [hunk 3]
```diff
@@ -2131,15 +2131,15 @@ int sdlog2_thread_main(int argc, char *argv[])
 			log_msg.body.log_ATT.q_x = buf.att.q[1];
 			log_msg.body.log_ATT.q_y = buf.att.q[2];
 			log_msg.body.log_ATT.q_z = buf.att.q[3];
-			log_msg.body.log_ATT.roll = buf.att.roll;
-			log_msg.body.log_ATT.pitch = buf.att.pitch;
-			log_msg.body.log_ATT.yaw = buf.att.yaw;
+			log_msg.body.log_ATT.roll = 0;
```
Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: A lot of the plotting tools don't support conversion. I would just convert here for logging.

File: src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp [hunk 4]
```diff
@@ -573,32 +573,32 @@ void AttitudeEstimatorQ::task_main()
 		struct vehicle_attitude_s att = {};
 		att.timestamp = sensors.timestamp;
 
-		att.roll = euler(0);
-		att.pitch = euler(1);
-		att.yaw = euler(2);
+		//att.roll = euler(0);
```
Comment (LorenzMeier) [inline src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp]: This should be removed, not commented.

File: src/examples/px4_simple_app/px4_simple_app.c [hunk 5]
```diff
@@ -107,9 +107,13 @@ int px4_simple_app_main(int argc, char *argv[])
 					 (double)raw.accelerometer_m_s2[2]);
 
 				/* set att and publish this information for other apps */
-				att.roll = raw.accelerometer_m_s2[0];
-				att.pitch = raw.accelerometer_m_s2[1];
-				att.yaw = raw.accelerometer_m_s2[2];
+				//att.roll = raw.accelerometer_m_s2[0];
```
Comment (LorenzMeier) [inline src/examples/px4_simple_app/px4_simple_app.c]: Remove, don't comment

File: cmake/configs/posix_sitl_default.cmake [hunk 6]
```diff
@@ -33,7 +33,7 @@ set(config_module_list
 	modules/sensors
 	modules/simulator
 	modules/mavlink
-	modules/attitude_estimator_ekf
+	#modules/attitude_estimator_ekf
```
Comment (LorenzMeier) [inline cmake/configs/posix_sitl_default.cmake]: We either should remove the code or not comment it. We need to build stuff that is in the codebase to ensure it doesn't bit-rot.

File: src/modules/ekf2/ekf2_main.cpp [hunk 7]
```diff
@@ -590,10 +590,10 @@ void Ekf2::task_main()
 
 			// generate remaining vehicle attitude data
 			att.timestamp = hrt_absolute_time();
-			matrix::Euler<float> euler(q);
-			att.roll = euler(0);
-			att.pitch = euler(1);
-			att.yaw = euler(2);
+			//matrix::Euler<float> euler(q);
```
Comment (LorenzMeier) [inline src/modules/ekf2/ekf2_main.cpp]: Remove, don't comment

File: src/examples/px4_simple_app/px4_simple_app.c [hunk 8]
```diff
@@ -107,9 +107,13 @@ int px4_simple_app_main(int argc, char *argv[])
...
+				//att.pitch = raw.accelerometer_m_s2[1];
+				//att.yaw = raw.accelerometer_m_s2[2];
+				att.q[0] = raw.accelerometer_m_s2[0];
+				att.q[1] = raw.accelerometer_m_s2[1];
+				att.q[2] = raw.accelerometer_m_s2[2];
+
```
Comment (kd0aij) [inline src/examples/px4_simple_app/px4_simple_app.c]: I assume this was misusing the roll/pitch/yaw fields to start with

**Comment by tumbili** [comment]
@LorenzMeier @kd0aij I fixed up the things you commented on. Just need to add a local quaternion to euler conversion into the sdlog2 because I can't use the math libraries. After that I'll give it a good test in SITL. Already tried quad and the quad plane in SITL today and they still seem to fly which was surprising.

File: src/modules/attitude_estimator_ekf/attitude_estimator_ekf_main.cpp [hunk 9]
```diff
@@ -605,34 +605,30 @@ int attitude_estimator_ekf_thread_main(int argc, char *argv[])
...
-					att.pitch = euler[1];
-					att.yaw = euler[2] + mag_decl;
 
 					att.rollspeed = x_aposteriori[0];
 					att.pitchspeed = x_aposteriori[1];
 					att.yawspeed = x_aposteriori[2];
-					att.rollacc = x_aposteriori[3];
-					att.pitchacc = x_aposteriori[4];
-					att.yawacc = x_aposteriori[5];
+					// att.rollacc = x_aposteriori[3];
```
Comment (LorenzMeier) [inline src/modules/attitude_estimator_ekf/attitude_estimator_ekf_main.cpp]: don't use comments

File: src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp [hunk 10]
```diff
@@ -573,32 +573,23 @@ void AttitudeEstimatorQ::task_main()
...
 		att.pitchspeed = _rates(1);
 		att.yawspeed = _rates(2);
 
-		for (int i = 0; i < 3; i++) {
-			att.g_comp[i] = _accel(i) - _pos_acc(i);
-		}
-
-		/* copy offsets */
-		memcpy(&att.rate_offsets, _gyro_bias.data, sizeof(att.rate_offsets));
+		//for (int i = 0; i < 3; i++) {
```
Comment (LorenzMeier) [inline src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp]: Don't comment out stuff

File: src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp [hunk 11]
```diff
@@ -573,32 +573,23 @@ void AttitudeEstimatorQ::task_main()
...
-		/* copy rotation matrix */
-		memcpy(&att.R[0], R.data, sizeof(att.R));
-		att.R_valid = true;
 		memcpy(&att.q[0], _q.data, sizeof(att.q));
 		att.q_valid = true;
 
-		att.rate_vibration = _voter_gyro.get_vibration_factor(hrt_absolute_time());
-		att.accel_vibration = _voter_accel.get_vibration_factor(hrt_absolute_time());
-		att.mag_vibration = _voter_mag.get_vibration_factor(hrt_absolute_time());
+		//att.rate_vibration = _voter_gyro.get_vibration_factor(hrt_absolute_time());
```
Comment (LorenzMeier) [inline src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp]: Don't comment out stuff

File: src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp [hunk 12]
```diff
@@ -827,37 +827,31 @@ void AttitudePositionEstimatorEKF::initializeGPS()
 void AttitudePositionEstimatorEKF::publishAttitude()
 {
 	// Output results
-	math::Quaternion q(_ekf->states[0], _ekf->states[1], _ekf->states[2], _ekf->states[3]);
-	math::Matrix<3, 3> R = q.to_dcm();
-	math::Vector<3> euler = R.to_euler();
+	matrix::Quaternion<float> q(_ekf->states[0], _ekf->states[1], _ekf->states[2], _ekf->states[3]);
+	//math::Matrix<3, 3> R = q.to_dcm();
```
Comment (LorenzMeier) [inline src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp]: Don't comment out stuff

File: src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp [hunk 13]
```diff
@@ -827,37 +827,31 @@ void AttitudePositionEstimatorEKF::initializeGPS()
...
+	//math::Vector<3> euler = R.to_euler();
 
-	for (int i = 0; i < 3; i++) {
+	/*for (int i = 0; i < 3; i++) {
```
Comment (LorenzMeier) [inline src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp]: Don't comment out stuff

File: src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp [hunk 14]
```diff
@@ -827,37 +827,31 @@ void AttitudePositionEstimatorEKF::initializeGPS()
...
 
 	_att.rollspeed = _ekf->dAngIMU.x / _ekf->dtIMU - _ekf->states[10] / _ekf->dtIMUfilt;
 	_att.pitchspeed = _ekf->dAngIMU.y / _ekf->dtIMU - _ekf->states[11] / _ekf->dtIMUfilt;
 	_att.yawspeed = _ekf->dAngIMU.z / _ekf->dtIMU - _ekf->states[12] / _ekf->dtIMUfilt;
 
 	// gyro offsets
-	_att.rate_offsets[0] = _ekf->states[10] / _ekf->dtIMUfilt;
-	_att.rate_offsets[1] = _ekf->states[11] / _ekf->dtIMUfilt;
-	_att.rate_offsets[2] = _ekf->states[12] / _ekf->dtIMUfilt;
+	//_att.rate_offsets[0] = _ekf->states[10] / _ekf->dtIMUfilt;
```
Comment (LorenzMeier) [inline src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp]: Don't comment out stuff

**Comment by tumbili** [comment]
@kd0aij Tried merging this branch into the new logger. When only logging attitude at 100Hz the log rate is at 3.5Kbytes/sec.

File: msg/vehicle_attitude.msg [hunk 15]
```diff
@@ -1,21 +1,7 @@
...
-float32 pitchacc	# Pitch angular acceleration (rad/s^2, x forward/y right/z down)
-float32 yawacc		# Yaw angular acceleration (rad/s^2, x forward/y right/z down)
-float32 rate_vibration	# Value between 0 and 1 indicating vibration. A value of 0 means no vibration, a value of 1 indicates unbearable vibration levels.
-float32 accel_vibration	# Value between 0 and 1 indicating vibration. A value of 0 means no vibration, a value of 1 indicates unbearable vibration levels.
-float32 mag_vibration	# Value between 0 and 1 indicating vibration. A value of 0 means no vibration, a value of 1 indicates unbearable vibration levels.
-float32[3] rate_offsets	# Offsets of the body angular rates from zero
-float32[9] R		# Rotation matrix, body to world, (Tait-Bryan, NED)
+float32 rollspeed	# Angular velocity about body north axis (x) in rad/s
+float32 pitchspeed	# Angular velocity about body east axis (y) in rad/s
+float32 yawspeed	# Angular velocity about body down axis (z) in rad/s
```
Comment (julianoes) [inline msg/vehicle_attitude.msg]: I vote to change these to `angular_rate_x`, `angular_rate_y`, `angular_rate_z`.

File: msg/vehicle_attitude.msg [hunk 16]
```diff
@@ -1,21 +1,7 @@
...
 float32[4] q		# Quaternion (NED)
-float32[3] g_comp	# Compensated gravity vector
-bool R_valid		# Rotation matrix valid
 bool q_valid		# Quaternion valid
```
Comment (julianoes) [inline msg/vehicle_attitude.msg]: If there is a `q_valid`, should we also have a `angular_rate_valid`, or is that one **always** valid?

File: src/examples/px4_simple_app/px4_simple_app.c [hunk 17]
```diff
@@ -106,10 +106,13 @@ int px4_simple_app_main(int argc, char *argv[])
...
-				/* set att and publish this information for other apps */
-				att.roll = raw.accelerometer_m_s2[0];
-				att.pitch = raw.accelerometer_m_s2[1];
-				att.yaw = raw.accelerometer_m_s2[2];
+				/* set att and publish this information for other apps
+				 the following does not have any meaning, it's just an example
+				*/
+				att.q[0] = raw.accelerometer_m_s2[0];
+				att.q[1] = raw.accelerometer_m_s2[1];
+				att.q[2] = raw.accelerometer_m_s2[2];
```
Comment (julianoes) [inline src/examples/px4_simple_app/px4_simple_app.c]: Ha, what random stuff!

File: src/modules/position_estimator_inav/position_estimator_inav_main.cpp [hunk 18]
```diff
@@ -1323,7 +1329,8 @@ int position_estimator_inav_thread_main(int argc, char *argv[])
 			local_pos.vy = y_est[1];
 			local_pos.z = z_est[0];
 			local_pos.vz = z_est[1];
-			local_pos.yaw = att.yaw;
+			matrix::Euler<float> euler(R);
```
Comment (julianoes) [inline src/modules/position_estimator_inav/position_estimator_inav_main.cpp]: What's cheaper: getting `euler(2)` from `q` or `R`?

File: src/modules/sdlog2/sdlog2.c [hunk 19]
```diff
@@ -2129,19 +2129,23 @@ int sdlog2_thread_main(int argc, char *argv[])
...
+			log_msg.body.log_ATT.roll = atan2f(2*(q0*q1 + q2*q3), 1 - 2*(q1*q1 + q2*q2));
+			log_msg.body.log_ATT.pitch = asinf(2*(q0*q2 - q3*q1));
+			log_msg.body.log_ATT.yaw = atan2f(2*(q0*q3 + q1*q2), 1 - 2*(q2*q2 + q3*q3));
 			log_msg.body.log_ATT.roll_rate = buf.att.rollspeed;
 			log_msg.body.log_ATT.pitch_rate = buf.att.pitchspeed;
 			log_msg.body.log_ATT.yaw_rate = buf.att.yawspeed;
-			log_msg.body.log_ATT.gx = buf.att.g_comp[0];
-			log_msg.body.log_ATT.gy = buf.att.g_comp[1];
-			log_msg.body.log_ATT.gz = buf.att.g_comp[2];
+			log_msg.body.log_ATT.gx = 0;
```
Comment (julianoes) [inline src/modules/sdlog2/sdlog2.c]: Why not remove this field?

File: src/modules/sdlog2/sdlog2.c [hunk 20]
```diff
@@ -2129,19 +2129,23 @@ int sdlog2_thread_main(int argc, char *argv[])
...
-			log_msg.body.log_ATT.yaw = buf.att.yaw;
+			float q0 = buf.att.q[0];
+			float q1 = buf.att.q[1];
+			float q2 = buf.att.q[2];
+			float q3 = buf.att.q[3];
+			log_msg.body.log_ATT.q_w = q0;
+			log_msg.body.log_ATT.q_x = q1;
+			log_msg.body.log_ATT.q_y = q2;
+			log_msg.body.log_ATT.q_z = q3;
+			log_msg.body.log_ATT.roll = atan2f(2*(q0*q1 + q2*q3), 1 - 2*(q1*q1 + q2*q2));
```
Comment (julianoes) [inline src/modules/sdlog2/sdlog2.c]: These should be the usual `euler(0)`, or not?

File: src/modules/vtol_att_control/standard.cpp [hunk 21]
```diff
@@ -300,29 +300,37 @@ void Standard::update_mc_state()
 
 	// get projection of thrust vector on body x axis. This is used to
 	// determine the desired forward acceleration which we want to achieve with the pusher
-	math::Matrix<3,3> R(&_v_att->R[0]);
-	math::Matrix<3,3> R_sp(&_v_att_sp->R_body[0]);
-	math::Vector<3> thrust_sp_axis(-R_sp(0,2), -R_sp(1,2), -R_sp(2,2));
-	math::Vector<3> euler = R.to_euler();
-	R.from_euler(0, 0, euler(2));
-	math::Vector<3> body_x_zero_tilt(R(0,0), R(1,0), R(2,0));
+	//math::Matrix<3,3> R(&_v_att->R[0]);
```
Comment (julianoes) [inline src/modules/vtol_att_control/standard.cpp]: Delete this line and add a comment instead.

File: src/modules/vtol_att_control/tailsitter.cpp [hunk 22]
```diff
@@ -131,6 +131,10 @@ void Tailsitter::update_vtol_state()
 	 * For the backtransition the pitch is controlled in MC mode again and switches to full MC control reaching the sufficient pitch angle.
 	*/
 
+	matrix::Quaternion<float> q(&_v_att->q[0]);
+	matrix::Euler<float> euler(q);
+	float pitch = euler(1);
```
Comment (julianoes) [inline src/modules/vtol_att_control/tailsitter.cpp]: If this is just another name for `euler(1)`, what about either:

`const float pitch = euler(1);`
or:
`float &pitch = euler(1);`

**Comment by julianoes** [comment]
Reviewed, didn't see anything stupid such as a typo.

**Comment by tumbili** [comment]
@julianoes Thanks for the review, I'll comment tomorrow.

File: src/modules/sdlog2/sdlog2.c (hunk 20 above)
Comment (tumbili) [inline src/modules/sdlog2/sdlog2.c]: @julianoes Those c++ functions are not available here.

File: src/examples/mc_att_control_multiplatform/mc_att_control_base.cpp [hunk 23]
```diff
@@ -87,11 +87,13 @@ void MulticopterAttitudeControlBase::control_attitude(float dt)
 
 	/* construct attitude setpoint rotation matrix */
 	math::Matrix<3, 3> R_sp;
-	R_sp.set(_v_att_sp->data().R_body);
+	matrix::Quaternion<float> q_sp(&_v_att_sp->data().q_d[0]);
+	R_sp.set(&q_sp._data[0][0]);
```
Comment (bkueng) [inline src/examples/mc_att_control_multiplatform/mc_att_control_base.cpp]: this seems wrong. it sets quaternion data to a matrix (below as well)

**Comment by tumbili** [comment]
@bkueng Thanks for the review! I'll give this a rebase in the evening and consider your comments.

**Comment by LorenzMeier** [comment]
@tumbili I will finish the current ones. But I think we should also fix the attitude setpoint accordingly. That's a similarly big high-rate topic that needs to be trimmed down to just the quaternion.

File: src/drivers/bst/bst.cpp [hunk 24]
```diff
@@ -289,11 +290,13 @@ void BST::cycle()
 		if (updated) {
 			vehicle_attitude_s att;
 			orb_copy(ORB_ID(vehicle_attitude), _attitude_sub, &att);
+			matrix::Quaternion<float> q(&att.q[0]);
```
Comment (jgoppert) [inline src/drivers/bst/bst.cpp]: I think that &att.q[0] and att.q are equivalent, both a pointer to the first entry, so might look better to type:

matrix::Quaternion<float> q(att.q);

Also, maybe more readable to use

using namespace matrix;

Quatf q(att.q);

File: src/drivers/bst/bst.cpp [hunk 25]
```diff
@@ -289,11 +290,13 @@ void BST::cycle()
...
+			matrix::Euler<float> euler(q);
 			BSTPacket<BSTAttitude> bst_att = {};
 			bst_att.type = 0x1E;
-			bst_att.payload.roll = swap_int32(att.roll * 10000);
-			bst_att.payload.pitch = swap_int32(att.pitch * 10000);
-			bst_att.payload.yaw = swap_int32(att.yaw * 10000);
+			bst_att.payload.roll = swap_int32(euler(0) * 10000);
```
Comment (jgoppert) [inline src/drivers/bst/bst.cpp]: You might consider euler.phi(), euler.theta(), euler.psi() accessors.

File: src/modules/commander/accelerometer_calibration.cpp [hunk 26]
```diff
@@ -693,8 +693,10 @@ int do_level_calibration(orb_advert_t *mavlink_log_pub) {
 		}
 
 		orb_copy(ORB_ID(vehicle_attitude), att_sub, &att);
-		roll_mean += att.roll;
-		pitch_mean += att.pitch;
+		matrix::Quaternion<float> q(&att.q[0]);
```
Comment (jgoppert) [inline src/modules/commander/accelerometer_calibration.cpp]: Eulerf euler = Quatf(att.q);

File: src/modules/commander/commander.cpp [hunk 27]
```diff
@@ -1165,7 +1173,9 @@ static void commander_set_home_position(orb_advert_t &homePub, home_position_s &
 	home.y = localPosition.y;
 	home.z = localPosition.z;
 
-	home.yaw = attitude.yaw;
+	matrix::Quaternion<float> q(&attitude.q[0]);
+	matrix::Euler<float> euler(q);
```
Comment (jgoppert) [inline src/modules/commander/commander.cpp]: Eulerf euler = Quatf(attitude.q);

File: src/modules/local_position_estimator/sensors/flow.cpp [hunk 28]
```diff
@@ -63,7 +63,10 @@ int BlockLocalPositionEstimator::flowMeasure(Vector<float, n_y_flow> &y)
 		return -1;
 	}
 
-	float d = agl() * cosf(_sub_att.get().roll) * cosf(_sub_att.get().pitch);
+	matrix::Quaternion<float> q(&_sub_att.get().q[0]);
+	matrix::Euler<float> euler(q);
```
Comment (jgoppert) [inline src/modules/local_position_estimator/sensors/flow.cpp]: Eulerf euler = Quatf(_sub_att.get().q);

File: src/examples/rover_steering_control/main.cpp [hunk 29]
```diff
@@ -178,7 +179,11 @@ void control_attitude(const struct vehicle_attitude_setpoint_s *att_sp, const st
 	/*
 	 * Calculate roll error and apply P gain
```
Comment (jgoppert) [inline src/examples/rover_steering_control/main.cpp]: Euler from DCM is cheapest.
 
  Euler(const Quaternion<Type> &q) :
        Vector<Type, 3>()
    {
        *this = Euler(Dcm<Type>(q));
    }

File: src/drivers/bst/bst.cpp (hunk 24 above)
Comment (jgoppert) [inline src/drivers/bst/bst.cpp]: Would be nice to convert sdlog2 to c++ at some point.

**Comment by LorenzMeier** [comment]
@tumbili Rebased and pushed. Its ready for a final review and test flights.

**Comment by tumbili** [comment]
@LorenzMeier Cool, thanks! I can give it a test flight tomorrow after doing a review.

**Comment by LorenzMeier** [comment]
@tumbili Your last two commits failed to compile pretty much everywhere. Did you test-compile locally?

**Comment by tumbili** [comment]
@LorenzMeier Yes I did actually. I'll check.

**Comment by tumbili** [comment]
@LorenzMeier Yes I did actually. I'll check.

**Comment by LorenzMeier** [comment]
I've tested this again in SITL, reviewed it again carefully and finally merged it.

**Pull request closed without merging**